### ISSUE TITLE
Feature Request: Get Cursor Icon Size

### ISSUE THREAD

**Issue by futile** [rust-windowing/winit#3788]
### Description

Hi,

I have a larger-than-usual cursor size, which causes tooltips to look like this:

![image](https://github.com/rust-windowing/winit/assets/1130337/377070f3-0f78-4e76-a16a-1df624a581f5)

I.e., the mouse cursor icon covers part of the cursor.

What I would like is a way to **get the dimensions of the current cursor icon**, so that I can adjust the tooltip's position.

I looked around a bit, and found https://github.com/rust-windowing/winit/issues/3306, which is about custom cursors, but it doesn't deal with cursor icon size.

For Windows, I found this (old, but might still work) post where someone figured out how to retrieve the cursor icon size:
https://forums.codeguru.com/showthread.php?449040-get-the-system-current-cursor-size&s=6da505d928b0363ffa51abafcdaefd0e&p=1757676#post1757676

I guess this is annoying to implement? :see_no_evil: Thanks for taking a look already! :)

### Relevant platforms

Windows, macOS, Wayland, X11, Web

**Comment by madsmtm** [comment]
Hmm, on macOS / AppKit, I don't think you can change the size of the cursor programmatically, and I also don't know of a way to query the size of it? This may not be information that's available on Wayland either.

There _is_ the concept of a "cursor hotspot" though (see [the docs](https://developer.apple.com/documentation/appkit/nscursor/1529096-hotspot)), I think that's available on some of the other platforms too, and will probably influence any calculation you would make here?

(I will note that another solution to this problem would be to make your "attack" button larger).

**Comment by futile** [comment]
Thanks for taking a look at this issue!

> (I will note that another solution to this problem would be to make your "attack" button larger).

Kind of, in this example I actually created a tooltip at the cursor position, because I wanted it to follow the cursor around. But yeah, if the tooltip was on the button, that would work.

Actually, I solved it by just moving the tooltip to the right of the button, which looks fine in my case (so very similar to what you proposed):

![image](https://github.com/rust-windowing/winit/assets/1130337/ec1ed814-eea0-4e8b-a898-52954ae05c34)

However, this is actually an issue in a *lot* of applications, e.g., in browsers:

![image](https://github.com/rust-windowing/winit/assets/1130337/d0a59eab-d287-4ebc-b08c-9abefaa70f42)

Which is why I'm kind of happy if it receives some attention :)

**Comment by kchibisov** [comment]
You can not use hotspot alone for that because you don't know the shape of the cursor. hotspot also solves literally nothing because it determines the top-left corner of the cursor icon relative to the pointer location, so this information can not solve the issue, because it shows the offset, and offset could be zero regardless of the size/shape  of the cursor icon.

However, if you have both size and hotspot you can approximate the area which will be obscured, but you need _both_ of them at the same time.

On Wayland neither are available for example, on macOS it's just hotspot. On X11 I'm pretty sure you can not do that unless somehow figure out what is actually used right now, but I'm pretty sure you can not do that, it's not reliable at all.

What I'd say that systems compositor should figure this stuff out, since they have all the information available in the end of the day.

We could still probably add a _best effort_ API, but it'll be really rare to have both hotspot and cursor icon size at the same time available, so the usefulness of such API is questionably since you need _both_.

**Comment by kchibisov** [comment]
I've kept only windows since it's likely old enough to have some _hacks_ in place or the API generally _stable enough_ for these hacks to work.

Also, if you set a _custom cursor_ you have all the sizes/hotspot informations available, so it's not an issue either.

**Comment by daxpedda** [comment]
Just stumbled on this for MacOS:
https://source.chromium.org/chromium/chromium/src/+/25cbdb930951c1b8b7863351c45c340fcc95fcc4:ui/base/cocoa/cursor_utils.mm;l=275-285;bpv=0;bpt=0

Basically there is a way to get the accessibility setting that determines cursor size, that way the native cursor size might be calculated. This is probably not really viable, but just leaving that here.

**Comment by notgull** [comment]
As far as I can tell there aren't any APIs in any eminent GUI frameworks that expose this functionality. I'm not even sure what such an API would look like.

**Comment by futile** [comment]
> As far as I can tell there aren't any APIs in any eminent GUI frameworks that expose this functionality. I'm not even sure what such an API would look like.

Yeah, the problem is really prevalent in basically all applications/OSs when increasing the cursor size (i.e., cursor will start noticeably overlapping tooltips). So I'm not really surprised that other GUI frameworks don't support it either :see_no_evil:
